applications/conceptual-model-editor/src/action/extend-selection-action.ts

```diff
@@ -0,0 +1,973 @@
+        selectionExtensionToExtend.selectionExtension.edgeSelection.push(edgeWhichAddedClass);
+        if(selectionExtensionToExtend.nodesToEdgesMapping[classId] === undefined) {
+            selectionExtensionToExtend.nodesToEdgesMapping[classId] = [];
+        }
+        selectionExtensionToExtend.nodesToEdgesMapping[classId].push(edgeWhichAddedClass);
+    }
+}
+
+const createEmptySelectionExtension = (): SelectionExtension => {
+    return {
+        selectionExtension: {
+            nodeSelection: [],
+            edgeSelection: [],
+        },
+        nodesToEdgesMapping: {},
+    };
+}
+
+
+export type NodeSelection = {
+    identifiers: string[],
+    areIdentifiersFromVisualModel: boolean,
+};
+
+
+/**
+ * Implementation note: Even when we get visual ids on input, we are working with the semantic ones,
+ * We get the correct form of id right at the end of the process, which is in the
+ * {@link addToExtensionIfSatisfiesVisibilityFilter} method.
+ * @param shouldExpandExternalModels If set to true, then allow the extension to go through external models -
+ * meaning through elements which are not expanded. If set to false then don't allow.
+ * Also If set to false, then this method this method is synchronous!!!
+ * @param semanticModelFilter Null if all models should be considered, otherwise record with modelID as key and
+ * true as value if the model should be considered, false if it shouldn't be.
+ * Models which are not part of the semanticModelFilter, are by default not considered.
+ * @param classesContext
+ * @param graph
+ * @returns Promise which contains the result after performing the extension corresponding to given arguments. The result Contains only the extension!
+ * The removal of duplicities and concat with the selection given on input is on the caller of the method.
+ * It is promise, because we allow to select classes which are in the external semantic model (SGOV) and not part of current model expansion.
+ * But this use-case is not accessible to user from editor, it is only accessible for internal usage in cme codebase, if {@link shouldExpandExternalModels} is set to true.
+ * The returned identifiers are of semantic entities. Only case when identifiers of visual entities are returned is when the property areIdentifiersFromVisualModel on {@link nodeSelection} equals true
+ * AND the visibility filter is set to "ONLY-VISIBLE" (it doesn't make sense otherwise, because in such case we are referencing entities which are not part of the visual model)
+ */
+export const extendSelectionAction = async (
+    notifications: UseNotificationServiceWriterType,
+    graph: ModelGraphContextType,
+    classesContext: ClassesContextType | null,
+    nodeSelection: NodeSelection,
+    extensionTypes: ExtensionType[],
+    visibilityFilter: VisibilityFilter,
+    shouldExpandExternalModels: boolean,
+    semanticModelFilter: Record<string, boolean> | null
+): Promise<SelectionExtension> => {
```

Comment:
The function being async is not ideal, it is explained in the comment - that it is async only if we pass in argument, that we want to perform expansion, that is search in external vocabularies (in SGOv for example), otherwise it is sync function. Ideally it would be 2 implementations sync/async.